src/ray/gcs/gcs_server/gcs_actor_distribution.h

```diff
@@ -0,0 +1,468 @@
+};
+
+/// `GcsJobDistribution` represents job distribution on the cluster nodes.
+/// It is responsible for the lifetime of all the `GcsJobSchedulingContext` and records
+/// the distribution of jobs on each node.
+class GcsJobDistribution {
+ public:
+  /// Create a `GcsJobDistribution`
+  ///
+  /// \param gcs_job_scheduling_factory Factory to create GcsJobSchedulingContext.
+  explicit GcsJobDistribution(
+      std::function<std::shared_ptr<GcsJobSchedulingContext>(const JobID &)>
+          gcs_job_scheduling_factory);
+
+  /// Add a actor worker assignment.
+  ///
+  /// \param actor_worker_assignment to be added.
+  bool AddActorWorkerAssignment(
+      std::shared_ptr<GcsActorWorkerAssignment> actor_worker_assignment);
+
+  /// Update the distribution of job on each node.
+  ///
+  /// \param actor_worker_assignment actor worker assignment which node id is updated.
+  bool UpdateNodeToJob(std::shared_ptr<GcsActorWorkerAssignment> actor_worker_assignment);
+
+  /// Get job scheduling context by the specified job id.
+  std::shared_ptr<GcsJobSchedulingContext> GetJobSchedulingContext(
+      const JobID &job_id) const;
+
+  /// Find or create a job scheduling context if not exist.
+  ///
+  /// \param job_id ID of the specified job.
+  /// \return job scheduling context found or created.
+  std::shared_ptr<GcsJobSchedulingContext> FindOrCreateJobSchedulingContext(
+      const JobID &job_id);
+
+  /// Remove actor worker assignments by the tuple(node_id, actor_worker_assignment_id,
+  /// job_id) and update the distribution of jobs on the node if needed.
+  ///
+  /// \param node_id ID of the specified node.
+  /// \param actor_worker_assignment_id ID of the gcs actor worker assignment to be
+  /// removed. \param job_id ID of the job related with the actor worker assignment to be
+  /// removed. \return actor worker assignment associated with the specified
+  /// tuple(node_id, actor_worker_assignment_id,job_id).
+  std::shared_ptr<GcsActorWorkerAssignment>
+  RemoveActorWorkerAssignmentByActorWorkerAssignmentID(
+      const NodeID &node_id, const UniqueID &actor_worker_assignment_id,
+      const JobID &job_id);
+
+  std::shared_ptr<GcsActorWorkerAssignment> GetActorWorkerAssignmentById(
+      const JobID &job_id, const UniqueID &actor_worker_assignment_id) const;
+
+ private:
+  void RemoveJobFromNode(const JobID &job_id, const NodeID &node_id);
+
+ private:
+  /// The distribution of jobs on each node.
+  absl::flat_hash_map<NodeID, absl::flat_hash_set<JobID>> node_to_jobs_;
+  /// Map from job id to job scheduling context.
+  absl::flat_hash_map<JobID, std::shared_ptr<GcsJobSchedulingContext>>
+      job_scheduling_contexts_;
+  /// Factory to create GcsJobSchedulingContext.
+  std::function<std::shared_ptr<GcsJobSchedulingContext>(const JobID &)>
+      gcs_job_scheduling_factory_;
+};
+
+/// GcsBasedActorScheduler implements a resource-based node selection.
+class GcsBasedActorScheduler : public GcsActorScheduler {
```

Comment:
What's the difference between GcsBasedActorScheduler and GcsActorScheduler?

Also, would nice to have comments on the basic strategy of this scheduler.

Is it possible to use composition instead of inheritance? I.e. GcsBasedActorScheduler implements GcsActorSchedulerInterface and owns an instance of GcsActorScheduler. It will result in better readability and testablility.

Reply:
I don't think composition is significantly better than inheritance in this case. There are only two inheritance layers here, and we only need to override two functions (`SelectNode` and `HandleWorkerLeaseReply`). Using composition has to delegate the majority of member functions.